x/nft/keeper/grpc_query.go

```diff
@@ -0,0 +1,191 @@
+	}
+
+	if len(request.ClassId) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "class id can not be empty")
+	}
+
+	owner, err := sdk.AccAddressFromBech32(request.Owner)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	balance := k.GetBalance(ctx, request.ClassId, owner)
+	return &nft.QueryBalanceResponse{Amount: balance}, nil
+}
+
+// Owner return the owner of the NFT based on its class and id, same as ownerOf in ERC721
+func (k Keeper) Owner(goCtx context.Context, request *nft.QueryOwnerRequest) (*nft.QueryOwnerResponse, error) {
+	if request == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "empty request")
+	}
+
+	if len(request.ClassId) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "class id can not be empty")
+	}
+	if len(request.Id) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "nft id can not be empty")
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	owner := k.GetOwner(ctx, request.ClassId, request.Id)
+	return &nft.QueryOwnerResponse{Owner: owner.String()}, nil
+}
+
+// Supply return the number of NFTs from the given class, same as totalSupply of ERC721.
+func (k Keeper) Supply(goCtx context.Context, request *nft.QuerySupplyRequest) (*nft.QuerySupplyResponse, error) {
+	if request == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "empty request")
+	}
+
+	if len(request.ClassId) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "class id can not be empty")
+	}
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	supply := k.GetTotalSupply(ctx, request.ClassId)
+	return &nft.QuerySupplyResponse{Amount: supply}, nil
+}
+
+// NFTsOfClass return all NFTs of a given class or optional owner, similar to tokenByIndex in ERC721Enumerable
+func (k Keeper) NFTsOfClass(goCtx context.Context, request *nft.QueryNFTsOfClassRequest) (*nft.QueryNFTsOfClassResponse, error) {
+	if request == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "empty request")
+	}
+
+	if len(request.ClassId) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "class id can not be empty")
+	}
+
+	var nfts []*nft.NFT
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// if owner is not empty, filter nft by owner
+	if len(request.Owner) > 0 {
+		owner, err := sdk.AccAddressFromBech32(request.Owner)
+		if err != nil {
+			return nil, err
+		}
+
+		ownerStore := k.getClassStoreByOwner(ctx, owner, request.ClassId)
+		pageRes, err := query.Paginate(ownerStore, request.Pagination, func(key []byte, _ []byte) error {
+			nft, has := k.GetNFT(ctx, request.ClassId, string(key))
+			if has {
+				nfts = append(nfts, &nft)
+			}
+			return nil
+		})
+
+		if err != nil {
+			return nil, err
+		}
+		return &nft.QueryNFTsOfClassResponse{
+			Nfts:       nfts,
+			Pagination: pageRes,
+		}, nil
+	}
+
+	nftStore := k.getNFTStore(ctx, request.ClassId)
+	pageRes, err := query.Paginate(nftStore, request.Pagination, func(_ []byte, value []byte) error {
+		var nft nft.NFT
+		k.cdc.MustUnmarshal(value, &nft)
```

Comment:
We are strongly discouraging using `Must*` in all functions which return or can return error.
We keep `MustUnmarshal` because some interface implementations, which don't return error, are using it. That was the main reason we didn't do it when we were refactoring codec interface last time.